### PR TITLE
[POAE7-2819] Expr Evaluator API

### PR DESCRIPTION
**Pull request by chaojun-zhang**
### What changes were proposed in this pull request?

Implementation of Expression Evaluator API

### Why are the changes needed?

Add Expression Evaluator API

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Need add unit test when TODO part is finished.
### Which label does this PR belong to?

Cider, publicApi

### CODE REVIEW

File: cpp/src/cider/exec/processor/ExpressionEvaluator.cpp [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+      std::make_shared<BatchProcessorContext>(context->getAllocator());
+  plan::SubstraitPlan substraitPlan{extendedExpression};
+  batchProcessor_ = cider::exec::processor::makeBatchProcessor(substraitPlan.getPlan(),
+                                                               batchProcessorContext);
+}
+
+std::unique_ptr<Batch> ExpressionEvaluator::evaluate(
+    const std::shared_ptr<Batch>& inBatch) {
+  batchProcessor_->processNextBatch(inBatch->getArray(), inBatch->getSchema());
+  batchProcessor_->finish();
```
Comment (Ma-Jian1) [inline cpp/src/cider/exec/processor/ExpressionEvaluator.cpp:44]: Can I call `evaluate` twice?

File: cpp/src/cider/include/cider/processor/ExpressionEvaluator.h [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+#include "cider/processor/BatchProcessor.h"
+#include "exec/nextgen/context/Batch.h"
+#include "substrait/algebra.pb.h"
+#include "substrait/extended_expression.pb.h"
+
+namespace cider::exec::processor {
+using ExpressionPtr = std::shared_ptr<::substrait::Expression>;
+using ExpressionVector = std::vector<ExpressionPtr>;
+
+using Batch = cider::exec::nextgen::context::Batch;
```
Comment (jikunshang) [inline cpp/src/cider/include/cider/processor/ExpressionEvaluator.h:35]: use `Batch` rather than `ArrowArray/ArrowSchema` as interface?

Comment (chaojun-zhang) [reply cpp/src/cider/include/cider/processor/ExpressionEvaluator.h:31]: Use batch would be more convenient than to use pair of ArrowArray/ArrowSchema. we should unify to use Batch instead of array and schema.

File: cpp/src/cider/exec/processor/ExpressionEvaluator.cpp (hunk 1 above)
Comment (chaojun-zhang) [reply cpp/src/cider/exec/processor/ExpressionEvaluator.cpp]: invoke 'finish' in destructor method now

File: cpp/src/cider/include/cider/processor/ExpressionEvaluator.h (hunk 2 above)
Comment (Ma-Jian1) [reply cpp/src/cider/include/cider/processor/ExpressionEvaluator.h:31]: users don't want to learn a new type, in the case of having `Arrow` already.

Comment (chaojun-zhang) [reply cpp/src/cider/include/cider/processor/ExpressionEvaluator.h:31]: I saw we use the new invent  type ' cider::exec::nextgen::context::Batch',  may be  use the [ArrowBatch ](https://arrow.apache.org/docs/cpp/api/table.html#_CPPv4N5arrow11RecordBatchE)   is make sense.

Comment (jikunshang) [reply cpp/src/cider/include/cider/processor/ExpressionEvaluator.h:31]: `RecordBatch` is C++ API. we should not use that.

Comment (chaojun-zhang) [reply cpp/src/cider/include/cider/processor/ExpressionEvaluator.h:31]: Any reason we don't want to use the C++ API instead of C ABI?

File: ci/docker/Dockerfile [hunk 3]
```diff
@@ -107,7 +107,7 @@ RUN mkdir -p ~/.gradle/ && \
 # @link: https://github.com/substrait-io/substrait-java/commit/5045aa3840156198d4251d4f3eb79440fc00e0e7
 RUN git clone https://github.com/substrait-io/substrait-java /substrait-java && \
     pushd /substrait-java && \
-    git checkout 5045aa3840156198d4251d4f3eb79440fc00e0e7 && \
+    git checkout 63cfcb33fa7818e91303d79317c12b8489a9014f && \
```
Comment (jikunshang) [inline ci/docker/Dockerfile:110]: please also update comments on L106

**Review by kyotoyx — APPROVED** [review]

File: .clang-tidy [hunk 4]
```diff
@@ -16,4 +16,4 @@
 # KIND, either express or implied.  See the License for the
 # specific language governing permissions and limitations
 # under the License.
-Checks: -*,readability-braces-around-statements,modernize-deprecated-headers,modernize-use-using,modernize-use-emplace,modernize-use-override
+Checks: -*,readability-braces-around-statements,modernize-deprecated-headers,modernize-use-using,modernize-use-emplace,modernize-use-override
```
Comment (yma11) [inline .clang-tidy:19]: revert this change?

File: cpp/src/cider/include/cider/processor/ExpressionEvaluator.h (hunk 2 above)
Comment (chaojun-zhang) [reply cpp/src/cider/include/cider/processor/ExpressionEvaluator.h:31]: updated

File: .clang-tidy (hunk 4 above)
Comment (chaojun-zhang) [reply .clang-tidy]: updated

File: ci/docker/Dockerfile (hunk 3 above)
Comment (chaojun-zhang) [reply ci/docker/Dockerfile:110]: updated

**Review by Ma-Jian1 — APPROVED** [review]
